### PR TITLE
Mongo

### PR DESCRIPTION
**Pull request by dingwenfang**

### CODE REVIEW

**Comment by caiwl** [comment]
@bill828 could you help to review?

**Comment by bill828** [comment]
> @bill828 could you help to review?



> @bill828 could you help to review?

sure, I'll review this PR today.

File: pkg/store/mongodb/mongodbStore.go [hunk 1]
```diff
@@ -0,0 +1,852 @@
...
+	matchstag := bson.D{{"$match", bson.D{{"$exists", true}}}}
+	projectstag := bson.D{{"$project", bson.D{{"_id", 1}}}}
+	opts := options.Aggregate().SetMaxTime(2 * time.Second)
+	cur, err := serviceCollection.Aggregate(ctx, mongo.Pipeline{matchstag, projectstag}, opts)
+	if err != nil {
+		return nil, err
+	}
+	var services []pms.Service
+	if err = cur.All(ctx, &services); err != nil {
+		panic(err)
```
Comment (bill828) [inline pkg/store/mongodb/mongodbStore.go:105]: Why panic here? panic may terminate the process. Is that possible to return the error?

File: pkg/store/mongodb/mongodbStore.go [hunk 2]
```diff
@@ -0,0 +1,852 @@
...
+
+}
+
+func (s *Store) DeletePolicies(serviceName string) error {
+	serviceCollection := s.client.Database(s.Database).Collection("services")
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+	filter := bson.D{{"_id", serviceName}}
+	update := bson.D{{"$pull", bson.D{{"policies", bson.D{{"$exists", true}}}}}}
+	result := serviceCollection.FindOneAndUpdate(ctx, filter, update)
```
Comment (bill828) [inline pkg/store/mongodb/mongodbStore.go:566]: Can result be null here?

File: pkg/store/mongodb/mongodbStore.go [hunk 3]
```diff
@@ -0,0 +1,852 @@
...
+	if result.Err() == mongo.ErrNoDocuments {
```
Comment (bill828) [inline pkg/store/mongodb/mongodbStore.go:567]: I believe the segment should be:
```
if result.Err() != nil {
    if result.Err() == mongo.ErrNoDocuments {
         return errors.Errorf(xxxxx)
    }
    return result.Err()
}
return nil
```

File: pkg/store/mongodb/mongodbStore.go [hunk 4]
```diff
@@ -0,0 +1,852 @@
...
+			}},
+		}}
+	opts := options.Aggregate().SetMaxTime(5 * time.Second)
+	cur, err := serviceCollection.Aggregate(ctx, mongo.Pipeline{matchstag, projectstag}, opts)
+	if err != nil {
+		return nil, err
+	}
+	var services []pms.Service
+	if err = cur.All(ctx, &services); err != nil {
+		panic(err)
```
Comment (bill828) [inline pkg/store/mongodb/mongodbStore.go:692]: Why panic here?

File: pkg/store/mongodb/mongodbStore_test.go [hunk 5]
```diff
@@ -0,0 +1,692 @@
...
+	}
+	if (counts["service1"].PolicyCount != int64(len(policies))) ||
+		(counts["service2"].PolicyCount != int64(len(policies))) {
+		t.Fatal("incorrect policy number")
+	}
+	if (counts["service1"].RolePolicyCount != int64(len(rolePolicies))) ||
+		(counts["service1"].RolePolicyCount != int64(len(rolePolicies))) {
+		t.Fatal("incorrect role policy number")
+	}
+	fmt.Println(counts)
```
Comment (bill828) [inline pkg/store/mongodb/mongodbStore_test.go:691]: uses `t.Log`.

File: pkg/store/mongodb/storeBuiler.go [hunk 6]
```diff
@@ -0,0 +1,83 @@
+//Copyright (c) 2018, Oracle and/or its affiliates. All rights reserved.
+//Licensed under the Universal Permissive License (UPL) Version 1.0 as shown at http://oss.oracle.com/licenses/upl.
+
+package mongodb
+
+import (
+	"context"
+
+	"github.com/prometheus/common/log"
```
Comment (bill828) [inline pkg/store/mongodb/storeBuiler.go:9]: We are using logrus, not prometheus log.

File: pkg/store/mongodb/storeBuiler.go [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+	// Set client options
+	clientOptions := options.Client().ApplyURI(mongoURI)
+	clientOptions.SetRetryReads(true)
+	clientOptions.SetRetryWrites(true)
+
+	// Connect to MongoDB
+	client, err := mongo.Connect(context.TODO(), clientOptions)
+
+	if err != nil {
+		log.Fatal(err)
```
Comment (bill828) [inline pkg/store/mongodb/storeBuiler.go:54]: Does `log.Fatal` terminate the process?

File: pkg/store/mongodb/mongodbStore.go (hunk 2 above)
Comment (dingwenfang) [reply pkg/store/mongodb/mongodbStore.go:566]: result will not be nil here

File: pkg/store/mongodb/mongodbStore.go (hunk 3 above)
Comment (dingwenfang) [reply pkg/store/mongodb/mongodbStore.go:567]: It seems no need to check if result.Err() != nil  here.

File: pkg/store/mongodb/mongodbStore.go (hunk 4 above)
Comment (dingwenfang) [reply pkg/store/mongodb/mongodbStore.go:692]: will fix it.

File: pkg/store/mongodb/mongodbStore_test.go (hunk 5 above)
Comment (dingwenfang) [reply pkg/store/mongodb/mongodbStore_test.go:691]: will fix it.

File: pkg/store/mongodb/storeBuiler.go (hunk 6 above)
Comment (dingwenfang) [reply pkg/store/mongodb/storeBuiler.go:9]: will fix it.

File: pkg/store/mongodb/storeBuiler.go (hunk 7 above)
Comment (dingwenfang) [reply pkg/store/mongodb/storeBuiler.go:54]: yes, will fix it.

File: pkg/store/mongodb/mongodbStore.go (hunk 1 above)
Comment (dingwenfang) [reply pkg/store/mongodb/mongodbStore.go:105]: will fix it.

File: pkg/store/mongodb/mongodbStore_test.go (hunk 5 above)
Comment (dingwenfang) [reply pkg/store/mongodb/mongodbStore_test.go]: will fix it

**Pull request merged**